File: src/53.cpp
class Solution {
public:
    int maxSubArray(vector<int>& nums) {
        if (nums.size() == 0) {
            return 0;
        } 
        if (nums.size() == 0) {
            return nums[0];
        }
        
        vector<int> dp(nums.size());
        dp[0] = nums[0];
        
        for (int i = 1; i < nums.size(); i ++) {
            dp[i] = max(dp[i - 1] + nums[i], nums[i]);
        }
        
        int max_val;
        max_val = INT_MIN;
        
        for (int i = 0; i < nums.size(); i ++) {
            if (max_val < dp[i]) {
                max_val = dp[i];
            }
        }
        
        return max_val;
    }
    
    int max(int a, int b) {
        if (a > b) {
            return a;
        } else {
            return b;
        }
    }
};

File: src/53.py
class Solution(object):
    def maxSubArray(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        if len(nums) == 0:
            return 0
        if len(nums) == 1:
            return nums[0]
        
        ##init
        dp = [0 for i in range(len(nums))]
        dp[0] = nums[0]
        
        for i in range(1, len(nums)):
            dp[i] = max(dp[i - 1] + nums[i], nums[i])
        
        return max(dp)
        
        ##
        #state function
        #dp表示i之前subarray的最大和
        #dp[i] = max(dp[i - 1] + num[i], num[i])
        
        #dp [-2,1,-2,4,3,5,6,1,4]
        